### PR TITLE
Added support for KakuteH7 flight controller

### PR DESCRIPTION
**Pull request by tridge**
H7 version of KakuteF7
Needs testing of all features

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+# like FPort
+PC7 USART6_RX USART6 NODMA ALT(1)
+PC6 USART6_TX USART6 NODMA
+
+# UART7 USED BY ESC FROM ORIGINAL DESIGN
+PE7 UART7_RX UART7
+PE8 UART7_TX UART7
+
+# Motors
+PB0  TIM1_CH2N TIM1 PWM(1) GPIO(50)
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat:115]: Could we make this board bi-directional dshot capable by default and use the non-complimentary timers?

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+# use last 2 pages for flash storage
+# H743 has 16 pages of 128k each
+define STORAGE_FLASH_PAGE 14
+define HAL_STORAGE_SIZE 32768
+
+# one I2C bus
+I2C_ORDER I2C1
+
+# order of UARTs (and USB)
+SERIAL_ORDER OTG1 USART1 USART2 USART3 UART4 UART7 USART6 OTG2
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat:31]: For what it's worth I think its much more understandable to have a 1-1 mapping of UART -> SERIAL, so just leave SERIAL5 empty and switch UART6 and UART7

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+PB1  TIM1_CH3N TIM1 PWM(2) GPIO(51)
+PB3  TIM2_CH2  TIM2 PWM(3) GPIO(52)
+PB10 TIM2_CH3  TIM2 PWM(4) GPIO(53)
+PA0  TIM5_CH1  TIM5 PWM(5) GPIO(54)
+PA2  TIM5_CH3  TIM5 PWM(6) GPIO(55)
+PC8  TIM3_CH3  TIM3 PWM(7) GPIO(56)
+PC9  TIM3_CH4  TIM3 PWM(8) GPIO(57)
+
+# LED strip
+PD12 TIM4_CH1 TIM4 PWM(9) GPIO(58)
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat:125]: I think it helps to include a defaults.parm file that setups the LED pin as NeoPixel

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat:115]: good idea. I've setup bidir on first 4 motors

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat (hunk 2 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat:31]: done

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat:115]: darn, I had to back this out. We can't support all 9 PWM and also support PPM on RC input if we do BIDIR on the first channels

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/hwdef.dat:125]: done

File: libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+## UART Mapping
+
+The UARTs are marked Rn and Tn in the above pinouts. The Rn pin is the
+receive pin for UARTn. The Tn pin is the transmit pin for UARTn.
+
+ - SERIAL0 -> USB
+ - SERIAL1 -> UART1 (Telem1)
+ - SERIAL2 -> UART2 (Telem2)
+ - SERIAL3 -> UART3 (GPS)
+ - SERIAL4 -> UART4
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/KakuteH7/README.md:30]: done

File: libraries/AP_HAL_ChibiOS/SPIDevice.cpp [hunk 5]
```diff
@@ -217,22 +217,39 @@ bool SPIDevice::do_transfer(const uint8_t *send, uint8_t *recv, uint32_t len)
...
         bus.semaphore.take_blocking();
         acquire_bus(true, true);
-        spiIgnore(spi_devices[device_desc.bus].driver, n);
+        osalSysLock();
+        spiStartIgnoreI(spi_devices[device_desc.bus].driver, n);
+        msg = osalThreadSuspendTimeoutS(&spi_devices[device_desc.bus].driver->thread, TIME_US2I(timeout_us));
+        osalSysUnlock();
+        if (msg == MSG_TIMEOUT) {
+            spiAbort(spi_devices[device_desc.bus].driver);
+        }
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/SPIDevice.cpp:237]: subroutine?

**Review by andyp1per — APPROVED** [review]

**Pull request merged**